File: src/converter/plugins/convertPropertyDeclaration.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {escapeIdentifier, isValidIdentifier} from "../../utils/strings.js";
import {renderNullable} from "../render.js";

export const convertPropertyDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isPropertyDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    checkCoverageService?.cover(node)

    const readonly = node.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)
    readonly && checkCoverageService?.cover(readonly)

    node.questionToken && checkCoverageService?.cover(node.questionToken)

    const modifier = readonly
        ? "val "
        : "var "

    let name: string

    if (ts.isStringLiteral(node.name)) {
        if (!isValidIdentifier(node.name.text)) {
            // TODO: generate inline getter
            return null
        }
        name = escapeIdentifier(node.name.text)
    } else if (ts.isNumericLiteral(node.name)) {
        // TODO: generate inline getter
        return null
    } else {
        name = escapeIdentifier(render(node.name))
    }

    const isOptional = Boolean(node.questionToken)

    const type = renderNullable(node.type, isOptional, context, render)

    return `${modifier}${name}: ${type}`
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertPropertyDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.escapeIdentifier
import typescript.SyntaxKind
import typescript.asArray
import typescript.isPropertyDeclaration

val convertPropertyDeclaration = createPlugin plugin@{ node, context, render ->
    if (!isPropertyDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService(checkCoverageServiceKey)

    checkCoverageService?.cover(node)

    val inheritanceModifierService = context.lookupService(inheritanceModifierServiceKey)

    val inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)

    val readonly = node.modifiers?.asArray()?.find { modifier -> modifier.kind === SyntaxKind.ReadonlyKeyword }
    readonly?.let { checkCoverageService?.cover(it) }

    node.questionToken?.let { checkCoverageService?.cover(it) }

    val modifier = if (readonly != null) "val " else "var "

    val name = escapeIdentifier(render(node.name))
    val annotation = createKebabAnnotation(node.name)

    val isOptional = node.questionToken != null

    val type = renderNullable(node.type, isOptional, context, render)

    "${ifPresent(annotation) { "${it}\n" }}${ifPresent(inheritanceModifier) { "$it "}}${modifier}${name}: $type"
}
